dashboard/src/shared/schema.ts

```diff
@@ -36,19 +36,24 @@ export function retrieveBasicFormParams(
           value,
           title,
           description,
+          children:
+            properties[propertyKey].type === "object"
+              ? retrieveBasicFormParams(defaultValues, properties[propertyKey], `${itemPath}.`)
+              : undefined,
         };
         params = {
           ...params,
           // The key of the param is the value of the form tag
           [form]: param,
         };
-      }
-      // If the property is an object, iterate recursively
-      if (schema.properties![propertyKey].type === "object") {
-        params = {
-          ...params,
-          ...retrieveBasicFormParams(defaultValues, properties[propertyKey], `${itemPath}.`),
-        };
+      } else {
+        // If the property is an object, iterate recursively
+        if (schema.properties![propertyKey].type === "object") {
+          params = {
+            ...params,
+            ...retrieveBasicFormParams(defaultValues, properties[propertyKey], `${itemPath}.`),
+          };
+        }
```

Comment:
So this is close, but not quite what I meant - it seems to enforce a hierarchy on the user (who defines the chart schema - that they include their database values in the structure we're wanting)? What I meant was that we, Kubeapps, can give the params the hierarchy we choose *after* we parse them.

So if this function did what it used to do - first producing a flat list of all the params identified in the schema - not caring about any hierarchy in the original file (as different charts can assume different things), then before returning that result *we* did a second pass (or call to separate function) to structure that data however we want (in this case, grouping all the ones which start with `externalDatabase`) and return those structured params.

This means that (a) we're not enforcing any structure on the end-user's values.yaml (well, schema), and (b) we no longer have the issue you mention above about the `USE_SELF_HOSTED_DB_PARAM_NAME` being treated differently, as we should group it with the same externalDatabase params (irrelevant where in the values/schema the user includes it) when returning our structured params.

Does that make sense? Let me know if I've misunderstood and it's not forcing a structure on the user as is.

Reply:
I see, I understand what you mean. What I like about having the hierarchy in the schema (or the values) is that we reduce the dependency between the keys used in the schema and the logic in Kubeapps. Right now it doesn't work this way but my plan is to remove the specific logic that checks that the `externalDatabaseHost` key is present (because it can be renamed to `externalDatabaseHostname`) and simply render as a subsection anything that is within the key `externalDatabase`.

I am okay doing the second round though to move the param `USE_SELF_HOSTED_DB_PARAM_NAME` as a children of `externalDatabase`, at least in that case we are adding a dependency with just one key name.

Reply:
Sounds like a good compromise :)